### PR TITLE
Add Paginator Functional Interface and ItemPaginator class

### PR DESCRIPTION
**Pull request by simpleauthority**
Add Paginator Functional Interface and ItemPaginator class for paginating lists of arbitrary data. Has to do with #3.

Y'all mind checking the code out too for style and to ensure it works well with ACF. :)

Here's how it looks when using ACF with it:
```
    @Subcommand("listflares|list|lf")
    @CommandPermission("flaredrops.listflares")
    public void listFlares(CommandSender sender, @Default("1") Integer page) {
        new ItemPaginator<>(
                "&eAvailable Flares",
                Flares.getFlares(),
                (flare) -> flare.getItem().getItemMeta().getDisplayName()
        ).sendPage(sender, page);
    }
```

Looks like so:
[
image
](url)

### CODE REVIEW

**Comment by simpleauthority** [comment]
Note to myself for when I have some time to look at this again:

- Remove the interface
- Rename ItemPaginator to Paginator
- Rethink page calculation, think preloading
- Cache a global paginator instance, or no? **Needs decision**
- To async, to sync _and_ async, or to sync only and let user async themselves? **Needs decision**
- Create fluent API for pagination (could be very cool if globally cached paginator is used)

@aikar, and who else should be tagged? :)

**Comment by simpleauthority** [comment]
Going to get some work done on this tonight, probably.

File: src/main/java/co/aikar/commands/ItemPaginator.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        }
+
+        page = Math.max(0, Math.min(page, maximumPages));
+
+        sender.sendMessage(formatHeader(header, (page + 1), (maximumPages + 1)));
+        for (int i = (PAGE_CAP * page); ((i < (PAGE_CAP * page) + PAGE_CAP) && i < items.size()); i++) {
+            sender.sendMessage(formatLine(stringMapper.apply(items.get(i))));
+        }
+    }
+}
```
Comment (kacperduras) [inline src/main/java/co/aikar/commands/ItemPaginator.java]: omomomomom

Comment (kacperduras) [inline src/main/java/co/aikar/commands/ItemPaginator.java]: omomomom.

**Comment by simpleauthority** [comment]
This is no longer relevant as we don't need a full interface for it. Closing.

**Pull request closed without merging**